crates/pixi_record/src/pinned_source.rs

```diff
@@ -129,6 +130,152 @@ impl PinnedSourceSpec {
+    ///     source: PinnedGitCheckout {
+    ///         commit: GitSha::from_str("abc123def456")?,
+    ///         subdirectory: None,
+    ///         reference: GitReference::DefaultBranch,
+    ///     },
+    /// });
+    ///
+    /// let source_spec = SourceSpec {
+    ///     location: SourceLocationSpec::Git(GitSpec {
+    ///         git: Url::parse("https://github.com/user/repo.git")?,
+    ///         rev: None,
+    ///         subdirectory: None,
+    ///     }),
+    /// };
+    ///
+    /// assert!(pinned_git.matches_source_spec(&source_spec));
+    /// # Ok(())
+    /// # }
+    /// ```
+    pub fn matches_source_spec(&self, source_spec: &SourceSpec) -> bool {
+        match (self, &source_spec.location) {
+            // Path sources: paths must be exactly equal
+            (PinnedSourceSpec::Path(pinned_path), SourceLocationSpec::Path(source_path)) => {
+                pinned_path.path == source_path.path
+            }
+
+            // Git sources: repository URLs must match, subdirectories must match if specified
+            (PinnedSourceSpec::Git(pinned_git), SourceLocationSpec::Git(source_git)) => {
+                use pixi_git::url::RepositoryUrl;
+
+                // Compare repository URLs (ignoring commit/branch details)
+                let pinned_repo = RepositoryUrl::new(&pinned_git.git);
+                let source_repo = RepositoryUrl::new(&source_git.git);
+
+                if pinned_repo != source_repo {
+                    return false;
+                }
+
+                // If source spec specifies a subdirectory, it must match
+                match (&source_git.subdirectory, &pinned_git.source.subdirectory) {
+                    (Some(source_subdir), Some(pinned_subdir)) => source_subdir == pinned_subdir,
+                    (Some(_), None) => false, // Source expects subdirectory, but pinned doesn't have one
+                    (None, _) => true,        // Source doesn't care about subdirectory
+                }
+            }
+
+            // URL sources: URLs must be exactly equal
+            (PinnedSourceSpec::Url(pinned_url), SourceLocationSpec::Url(source_url)) => {
+                pinned_url.url == source_url.url
+            }
+
+            // Mismatched types never match
+            _ => false,
+        }
+    }
+
+    /// Makes this pinned source relative to another pinned source if both are path sources
+    /// or both are git sources pointing to the same repository.
+    /// This is useful for making `build_source` relative to `manifest_source` in lock files.
+    ///
+    /// Returns `None` if:
+    /// - Not a compatible combination (different types or different git repos)
+    /// - The sources cannot be made relative to each other
+    ///
+    /// # Arguments
+    /// * `base` - The base pinned source to make this path relative to (typically the manifest_source)
+    pub fn make_relative_to(&self, base: &PinnedSourceSpec, workspace_root: &Path) -> Option<Self> {
```

Comment:
Maybe this should return a relative path instead of `Self`. The function only ever returns `None` or a relative path.